articles/150_roslaunch.md

```diff
@@ -428,62 +428,179 @@ The remapping design document[^static_remapping] goes into detail on how you can
+However, there also is no way to get feedback about the success or failure of loaded nodes.
+STDOUT cannot be used because a composable node logging messages to STDOUT is assumed to be very common and would conflict.
+Since STDIN is always available, it would be possible to unload a node via this API.
+
+###### API using ROS Services or Topics
+
+Lastly, a container process API may be defined by ROS services or topics.
+
+Advantages
+* No config file parsing
+* Works with read-only file systems
+* Can indicate if a node was successfully loaded
+* Can create API to trigger launch events
+* Can tell if a container process supports this interface
+
+Disadvantages
+* Must wait for the service API to become available
+* Cannot stop dynamically loaded nodes from creating the same services
+
+This is the only option discussed which can communicate the success or failure of dynamically launched nodes.
+It is also the only option that allows introspection.
+However, this option has the highest potential delay from when the container process is spawned to when nodes may be loaded.
+
+##### Proposed Container process API
+
+This is a proposal for a container process API to be used by the launch system.
+
+###### Command Line Arguments
+The container process must accept command line arguments including log level, remapping, and parameters.
+These command line arguments must not be applied to dynamically launched nodes.
+The launch system will pass these arguments to the container process in the same way it would pass them to a node.
+If a remap rule would apply to a launch service, the launch system should try to use the remapped service name instead.
+
+###### ROS Services
+The container process must offer all of the following services.
 
-</node_container_process>
+* `_launch/load_node`
+* `_launch/unload_node`
+* `_launch/list_nodes`
+
+The services are hidden to avoid colliding with user created services.
+`load_node` will be called by the launch system when a composable node is to be dynamically loaded, and `unload_node` destroys a composable node.
+`list_nodes` is not called by launch system, and is only provided for introspection.
+
+####### load_node
+
+If a container process is asked to load a node with a full node name matching an existing node, then it must reject the request.
+This is to avoid conflicts in features that assume node name uniqueness, like parameters.
+
+A container process must assign the node a unique id when it is loaded.
+The id of a loaded node instance never changes.
+Two nodes in the same container process must never have the same id, and there should be a significant time delay before an id is reused.
+
+```
+# A ROS package the composable node can be found in
+string package_name
+# a plugin within that package
+string plugin_name
+
+# Name the composable node should use, or empty to use the node's default name
+string node_name
+# Namespace the composable node should use, or empty to use the node's default namespace
```

Comment:
What if the node has a default namespace but the caller want's to instantiate it without a namespace?

Reply:
Yeah, I wasn't sure if this would be absolute or relative. Passing `/` would indeed work.